src/cluster.c

```diff
@@ -6479,6 +6485,16 @@ void removeChannelsInSlot(unsigned int slot) {
     zfree(channels);
 }
 
+/* Remove all the shard channels related information on the current node. */
+void removeAllShardChannelSubscription() {
+    clusterNode *currmaster = nodeIsMaster(myself) ? myself : myself->slaveof;
+    for (int j = 0; j < CLUSTER_SLOTS; j++) {
+        if (server.cluster->slots[j] == currmaster) {
+            removeChannelsInSlot(j);
```

Comment:
I wonder if it would be more defensive to do a check and call `removeChannelsInSlot` for all slots not-owned by the new master. We have a pretty long history of trying to handle these state transitions correctly, and missing edge cases like a slot was migrated and during some edge case it was missed and missed evicting a slot here. It would likely be slower, but I'm not sure I will worry too much about the performance for the promotion case. Maybe there is some smart way to optimize it.

Reply:
I think this method would become confusing with the above suggestion. I would rather remove the condition and call `removeChannelsInSlot` for each slot and do a complete cleanup.

Reply:
I like this suggestion (change == to !=). We would need to rename this function to sth like `removeAllNotOwnedShardChannelSubscriptions`. Then it would not be confusing, would it?

Then there's no need for the memcmp on shard id before calling it.

Reply:
I mean this method seems weird to be invoked from any other place. However, I see the gain in this case.

Reply:
Moving the invocation of this method after the master ownership transfer.